include/deal.II/base/std_cxx14/memory.h

```diff
@@ -30,15 +30,39 @@ namespace std_cxx14
 #ifdef DEAL_II_WITH_CXX14
   using std::make_unique;
 #else
+  namespace internal
+  {
+    template <typename T>
+    struct is_bounded_array
+    {
+      static constexpr bool value = false;
+    };
+
+    template <typename T, std::size_t N>
+    struct is_bounded_array<T[N]>
+    {
+      static constexpr bool value = true;
+    };
+  }
+
   template <typename T, typename... Args>
   inline
-  std::unique_ptr<T>
+  typename std::enable_if<!std::is_array<T>::value, std::unique_ptr<T> >::type
   make_unique(Args &&... constructor_arguments)
   {
-    static_assert(!std::is_array<T>::value,
-                  "This function is not implemented for array types.");
     return std::unique_ptr<T>(new T(std::forward<Args>(constructor_arguments)...));
   }
+
+  template <typename T>
+  inline
+  typename std::enable_if<std::is_array<T>::value, std::unique_ptr<T> >::type
+  make_unique(std::size_t n)
+  {
+    static_assert(!internal::is_bounded_array<T>::value,
+                  "This function is not implemented for bounded array types.");
+    return std::unique_ptr<T>(new typename std::remove_extent<T>::type [n]);
```

Comment:
why won't this work with fixed array lengths?

Reply:
As can be seen in https://godbolt.org/g/n6TsgS, `std::unique_ptr<int[7]>` cannot be initialized by a `int *` object since these are different types. The problem is that a `int[7]` variable immediately decays to a `int*`. 
In the end, the standard wants this overload to be deleted anyway. A `static_assert` just seemed easier.

Reply:
The reason why C++ doesn't allow fixed-sized arrays as argument types is that there is no way to initialize them. You can't pass in an array of size 7 as argument to the function -- or better, you can, but you can't copy a whole array. And you also can't pass in 7 separate arguments because they can't be used to copy-construct the elements of an array. In general, initializing arrays of any size that are allocated on the heap is not possible with the exception of initializing them all to the same value.

```
  std::make_shared<int[]>(7);
```
as in the goldbolt example does something different: it allocates an integer array of length 7. But it does not initialize the 7 elements to any given initializer.

Reply:
This is such a counterintuitive part of the language. I can't figure out how to initialize

```cpp
std::unique_ptr<int[5]> x;
```
and I am not so sure that there is even an answer.

Reply:
You can't initialize it. But that's no different from the following:
```
  typedef int five_ints[5];
  five_ints *p = new five_ints;
```
You can allocate such an object, but there is no syntax to initialize the 5 values at the same time as you allocate them, except that
```
  five_ints *p = new five_ints();
```
calls the default constructor for each integer (which sets them to zero).